File: src/common/py/api/authorization/authorization_commands.py
import typing
from dataclasses import field

from ...core.messaging import Message, Command, CommandReply
from ...core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)
from ...data.entities.authorization import AuthorizationToken
from ...data.entities.user import UserID
from ...integration.authorization import AuthorizationRequestPayload


@Message.define("command/authorization/request")
class RequestAuthorizationCommand(Command):
    """
    Command to perform an authorization request.

    Notes:
        Requires a ``RequestAuthorizationReply`` reply.

    Args:
        request_payload: The authorization request information.
        strategy: The token strategy (e.g., OAuth2).
        data: The actual token request data.
    """

    request_payload: AuthorizationRequestPayload = field(
        default_factory=AuthorizationRequestPayload
    )
    strategy: str
    data: typing.Dict[str, typing.Any] = field(default_factory=dict)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        request_payload: AuthorizationRequestPayload,
        strategy: str,
        data: typing.Dict[str, typing.Any],
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            RequestAuthorizationCommand,
            chain,
            request_payload=request_payload,
            strategy=strategy,
            data=data,
        )


@Message.define("command/authorization/request/reply")
class RequestAuthorizationReply(CommandReply):
    """
    Reply to ``RequestAuthorizationCommand``.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: RequestAuthorizationCommand,
        *,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            RequestAuthorizationReply, cmd, success, message
        )


@Message.define("command/authorization/revoke")
class RevokeAuthorizationCommand(Command):
    """
    Command to revoke an authorization.

    Notes:
        Requires a ``RevokeAuthorizationReply`` reply.

    Args:
        user_id: The user ID.
        auth_id: The ID of the token to revoke.
        force: If true, the token will be removed immediately; otherwise, it will be marked as invalid only
    """

    user_id: UserID
    auth_id: str

    force: bool

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        user_id: UserID,
        auth_id: str,
        force: bool = True,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            RevokeAuthorizationCommand,
            chain,
            user_id=user_id,
            auth_id=auth_id,
            force=force,
        )


@Message.define("command/authorization/revoke/reply")
class RevokeAuthorizationReply(CommandReply):
    """
    Reply to ``RevokeAuthorizationCommand``.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: RevokeAuthorizationCommand,
        *,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            RevokeAuthorizationReply, cmd, success, message
        )


@Message.define("command/authorization/token/get", is_protected=True)
class GetAuthorizationTokenCommand(Command):
    """
    Command to retrieve an authorization token.

    Notes:
        This message is protected and requires an API key. Requires a ``GetAuthorizationTokenReply`` reply.

    Args:
        user_id: The user ID.
        auth_id: The ID of the token to revoke.
    """

    user_id: UserID
    auth_id: str

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        user_id: UserID,
        auth_id: str,
        api_key: str,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            GetAuthorizationTokenCommand,
            chain,
            user_id=user_id,
            auth_id=auth_id,
            api_key=api_key,
        )


@Message.define("command/authorization/token/get/reply", is_protected=True)
class GetAuthorizationTokenReply(CommandReply):
    """
    Reply to ``GetAuthorizationTokenCommand``.

    Notes:
        This message is protected and requires an API key.
    """

    auth_token: AuthorizationToken | None

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: GetAuthorizationTokenCommand,
        *,
        auth_token: AuthorizationToken | None,
        api_key: str,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            GetAuthorizationTokenReply,
            cmd,
            success,
            message,
            auth_token=auth_token,
            api_key=api_key,
        )

File: src/common/web/api/authorization/AuthorizationCommands.ts
import { Command } from "../../core/messaging/Command";
import { CommandReply } from "../../core/messaging/CommandReply";
import { CommandComposer } from "../../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Message } from "../../core/messaging/Message";
import { AuthorizationTokenType } from "../../data/entities/authorization/AuthorizationToken";
import { type UserID } from "../../data/entities/user/User";
import { type AuthorizationRequestPayload } from "../../integration/authorization/AuthorizationRequest";

/**
 * Command to perform an authorization request. Requires a ``RequestAuthorizationReply`` reply.
 *
 * @param payload - The authorization request information.
 * @param strategy - The token strategy (e.g., OAuth2).
 * @param data - The actual token request data.
 */
@Message.define("command/authorization/request")
export class RequestAuthorizationCommand extends Command {
    public readonly request_payload: AuthorizationRequestPayload = {
        auth_id: "",
        auth_type: AuthorizationTokenType.Invalid,
        auth_issuer: "",
        auth_bearer: "",
        fingerprint: ""
    };
    public readonly strategy: string = "";
    public readonly data: any = null;

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        request_payload: AuthorizationRequestPayload,
        strategy: string,
        data: any,
        chain: Message | null = null
    ): CommandComposer<RequestAuthorizationCommand> {
        return messageBuilder.buildCommand(
            RequestAuthorizationCommand,
            {
                request_payload: request_payload,
                strategy: strategy,
                data: data
            },
            chain
        );
    }
}

/**
 * Reply to ``RequestAuthorizationCommand``.
 */
@Message.define("command/authorization/request/reply")
export class RequestAuthorizationReply extends CommandReply {
    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: RequestAuthorizationCommand,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<RequestAuthorizationReply> {
        return messageBuilder.buildCommandReply(RequestAuthorizationReply, cmd, success, message);
    }
}

/**
 * Command to revoke an authorization. Requires a ``RevokeAuthorizationReply`` reply.
 *
 * @param user_id - The user ID.
 * @param auth_id - The ID of the token to revoke.
 * @param force - If true, the token will be removed immediately; otherwise, it will be marked as invalid only
 */
@Message.define("command/authorization/revoke")
export class RevokeAuthorizationCommand extends Command {
    public readonly user_id: UserID = "";
    public readonly auth_id: string = "";

    public readonly force: boolean = true;

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        userID: UserID,
        authID: string,
        force: boolean = true,
        chain: Message | null = null
    ): CommandComposer<RevokeAuthorizationCommand> {
        return messageBuilder.buildCommand(RevokeAuthorizationCommand, { user_id: userID, auth_id: authID, force: force }, chain);
    }
}

/**
 * Reply to ``RevokeAuthorizationCommand``.
 */
@Message.define("command/authorization/revoke/reply")
export class RevokeAuthorizationReply extends CommandReply {
    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: RevokeAuthorizationCommand,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<RevokeAuthorizationReply> {
        return messageBuilder.buildCommandReply(RevokeAuthorizationReply, cmd, success, message);
    }
}
